pkg/loop/internal/relayer/pluginprovider/chainreader/chain_reader_test.go

```diff
@@ -524,11 +514,57 @@ func (f *fakeChainReader) QueryKey(_ context.Context, _ string, filter query.Key
 			})
 		}
 
-		return sequences, nil
+		// simulate contract being bound at a certain block not having the events before that block unless Replay is called.
+		boundAt, ok := f.contractBoundAt[contractName]
+		if !ok {
+			return nil, fmt.Errorf("contract %s not bound", contractName)
+		}
+
+		if boundAt == 0 {
+			return sequences, nil
+		}
+
+		return sequences[:boundAt], nil
 	}
 	return nil, nil
 }
 
+func (f *fakeChainReader) ReplaySequence(_ context.Context, contractName, _ string, blockID string) error {
+	blockNumber, err := strconv.Atoi(blockID)
+	if err != nil {
+		return err
+	}
+
+	if blockNumber > len(f.triggers) {
+		return fmt.Errorf("blockID: %d requested is in the future: %d", blockNumber, len(f.triggers))
+	}
+
+	// we only have a single event type used in tests for querying, so we can ignore key and just use contract name.
+	f.contractBoundAt[contractName] = blockNumber
```

Comment:
why no write lock?

Reply:
Tests are not parallel, only encoding tests are, but they don't reuse the tester instance

Reply:
why do all the other methods lock then? it seems inconsistent. should they be removed?

are you saying that the tests must, by some inherent characteristic, be serial or that the current test happen to be written that way?

Reply:
Initially this was testable in parallel, but the tests got a lot more complicated and share a lot of state. Also I am not sure if this can even work in parallel since then we are enforcing all of relayer implementations to also adhere to that. Anwyay I created a ticket to revisit this to see if it makes sense to work in parallel and if it does to make fix this by fixing the shared state and potential race conditions.[ BCF-3327](https://smartcontract-it.atlassian.net/browse/BCF-3327)